feature: added custom fps video capture setting

s0llvan opened the pull request:
I used the CaptureRequest.Builder object and also MediaRecorder.
MediaRecorder is apparently not compatible with all devices and CaptureRequest too, so as a precaution I used both ways for greater devices coverage.

I put four options : 24, 30, 60 and 90 (maybe 24 fps is not necessary ?)

I couldn't test the video capture at 60 and 90 fps because I don't have the necessary hardware but 24 and 30 fps modes work well.
See if mode 60 and 90 fps changes file size (which does not seem to be the case with 24 and 30 fps modes). Otherwise the storage space information widget will also need to be updated.

![Screenshot_20240916_210441](https://github.com/user-attachments/assets/1890941a-c1fa-4426-9a2a-432ddeb9a0d7)

I have also grouped constants used in several files into the same Constants file to avoid duplication.

#23

anonfaded commented:
> CaptureRequest.Builder

I'm not much familiar with android dev, so can you please explain about what this actually is?

> See if mode 60 and 90 fps changes file size (which does not seem to be the case with 24 and 30 fps modes). Otherwise the storage space information widget will also need to be updated.

I will try this later i think my phone supports 60 fps, but i think if you use logging then you can check in the `logcat` the selected fps. 
And the storage widget updates according to the file size automatically so we likely don't need to change anything in that. For reference you can check that yourself by recording video in sd quality and then fhd or hd where you can see the change size ratio.

And yeah i think 30 fps minimum is enough, but for now leave it as it is as first we need to check if 60fps works or not, if it doesn't show difference on my phone due to no hardware capability then we can ask other people on discord if they can try it.

s0llvan commented:
Indeed, i'm made a mistake about storage widget.

CaptureRequest is used for video capture and MediaRecorder for recording, but the problem is that limiting FPS only on recording is not compatible with all devices so we also need to limit FPS to capturing video (and limiting FPS on video capture is also not compatible with all devices so we also need to limit FPS on recording).

Sorry, but I don't understand how to test FPS with logcat 😅.
I tried with a virtual device but I can't exceed the 8 fps 😬.

anonfaded commented:
> Indeed, i'm made a mistake about storage widget.
> 
> CaptureRequest is used for video capture and MediaRecorder for recording, but the problem is that limiting FPS only on recording is not compatible with all devices so we also need to limit FPS to capturing video (and limiting FPS on video capture is also not compatible with all devices so we also need to limit FPS on recording).
> 
> Sorry, but I don't understand how to test FPS with logcat 😅.
> I tried with a virtual device but I can't exceed the 8 fps 😬.

Okay.
By logcat i just meant like if you make debugging statements like i have done for other methods where we print in the logcat (logcat is like a terminal in Android studio), like if you search in codebase Log.d then for example you will see lines like: Log.d(tag here, message here). So if you add such logging statements and print the selected fps then it may indicate if it's selected or not. But ok if you don't do it too, first I'll check if it works on my device. Until then if you try it on your physical device, by changing fps and then seeing in the gallery if it shows the fps(you need to save video to gallery after recording, by default it will only be saved in the internal storage of app which you will see in that records page) then it might indicate that too.

I'll merge this PR and then try it now.

s0llvan commented:
Yes I tested in 24 fps and 30 fps and it works. I recorded the videos in the gallery and sent them to my computer to check the framerate and every time I have the right maximum FPS 👌

anonfaded commented:
> Yes I tested in 24 fps and 30 fps and it works. I recorded the videos in the gallery and sent them to my computer to check the framerate and every time I have the right maximum FPS 👌

Btw i see you have removed the bitrates from the quality selection feature. But those were linked with the quality options like for sd it was lesser while for fhd it was higher, and now as you removed those bitrates won't it break the quality selection feature? Have you tried that?

Or maybe i misunderstood something?

s0llvan commented:
I moved the setVideoFrameRate function because it's duplicated but I don't think I've removed the bitrates ?

At this section or somewhere else ?

![image](https://github.com/user-attachments/assets/f82b930f-c266-4d75-b771-8d02cda02573)

anonfaded commented:
letme merge and then check as rn its hard haha

anonfaded commented:
okay so i tried it, and i see that even after i select 60 or 90 fps, it still records in 30fps maximum, while on 24 it records in 24 correctly, so here i'm not sure if this is bcz of incompatible device or some bug in the feature.

So for this is it possible to show a pop up message if a framerate isn't compatible with the device? like if user selects 60 fps while its not supported then they get a message that its not supported? If this happens then it will be very much better otherwise the user won't know what's happening actually.

Also right now we are not sure if the 60 and 90 fps works or not so i will ask a user and will let you know.

s0llvan commented:
Thank you for your feedback.

Indeed, I think a pop up message indicating an incompatibility will be useful, I had partly done it locally as well for the tests. I'll take care of it as soon as possible !

You can also check normally on the internet if video capture of your phone is compatible with 60 fps or 90 fps. But if you find someone it should work for, it's perfect 🙂

anonfaded commented:
Yeah i just checked online and my device is 30fps. Overall the feature seems fine so we can now wait for feedback from users, especially the person who asked for this feature.

Once again thanks, this was a great contribution! 😃 
I have to add more features and then new release will be published in some days. Will add a `credits` section in about page, onboarding on first app launch and other stuff.

Also there is a similar issue where i think we need to add a similar `bitrate selection` feature as a user said the audio quality isnt good right now(https://github.com/anonfaded/FadCam/issues/29). Just sharing if it interests you.

s0llvan commented:
@anonfaded

I wonder if the message is useful.
Couldn't we simply display only the FPS list compatible with his phone ?
